### ISSUE TITLE
Objectif final de UMAP-clustering

### ISSUE THREAD

**Issue by mateomorin** [victorgalmiche/umap-clustering-mise-en-prod#21]
## Contexte

On est encore à un stade où on peut encore décider précisément de la démarche à suivre, donc autant se poser les bonnes questions maintenant que dans une ou deux semaines. La question est : quel est l'objectif a-t-on derrière ce repo ? Reformulé autrement : **quelle contribution souhaitons-nous apporter ?**

## Base de travail

Le projet initial repose sur l'analyse de UMAP allié au clustering pour des données en grande dimension. Alors, on y retrouve : 
- un algorithme UMAP personnel : calcul exact (KDTree de sklearn) ou approximatif (NNDescent fait main par @victorgalmiche) des plus proches voisins, création d'un graphe avec des poids adaptés, projection en basse dimension grâce au Lagrangien, et optimisation des positions par des jeux de forces attractives/répulsives (réalisé par moi).
- la possibilité de visualiser en 2D l'évolution des embeddings au cours des itérations de l'algorithme UMAP.
- des tests de UMAP sur des données diverses, couplé à du clustering HDBSCAN/KMeans, surtout réalisés par @paquitopg, qui ont montré de très bonnes performances en 2D, en plus de faire de beaux visuels.

Gros soucis auquel on s'est confrontés : l'algorithme UMAP personnel a du mal à être robuste et la qualité du résultat dépend fortement de la donnée en entrée. Notamment, la partie sur la recherche des plus petites valeurs singulières du Laplacien, ou la recherche des paramètres pour la création des poids (bisect) peut ne pas converger. C'est pour ça qu'on a souvent fait appel à la librairie `umap-learn` pour avoir des données cohérentes à étudier.

**Micro-conclusion** : on a pu montrer la pertinence du clustering avec UMAP, mais notre algorithme personnel ne peut pas en l'état être robuste en production.

## Mon point de vue sur les contributions à apporter

Ce que je me dis :
1. Si l'algotihme fait main n'est pas assez robuste, on ne peut pas le manipuler à tout va dans tous les contextes.
2. La librairie `umap-learn` produit des résultats robustes, quoique parfois après de longs temps de calcul.
3. On veut montrer la puissance de l'alliance UMAP + clustering pour des jeux de données diverses.
4. On veut proposer un résultat visuel

Alors, on met l'accent sur la contribution principale (qui était aussi celle du papier qu'on avait publié au S1) : UMAP et clustering c'est vraiment robuste et ça produit des jolis résultats. Pour ce qui est du reste des contributions (algorithme), on fait au mieux mais sinon on ne fait pas.

## Comment faire ?

J'opterais donc personnellement pour une app (API/site web) qui, pour chaque donnée, renvoie un embedding 2D et un label d'appartenance à un cluster à partir d'un dataset entier en entrée. Les étapes principales du process : 
1. L'utilisateur envoie son jeu de données (CSV ou Parquet), ou prend un jeu d'exemple (utile pour les checks rapides).
2. On limite la donnée aux 500 première lignes (on veut faire de la visualisation notre priorité, et UMAP prend énormément de temps à run lorsque la dimension et la taille augmente).
3. On run UMAP pour différents `n_neighbors`, `epochs`, `learning_rate` et `KNN_metric` (à mettre dans la config). On essaie d'abord avec le UMAP fait main, si ça plante on fait avec le UMAP de `umap-learn`. On retient la trustworthiness.
4. On run HDBSCAN sur tous/une sélection des résultats (avec potentiellement différents paramètres), on retient la silouhette.
5. On fait une note générale pour chaque combinaison et on retourne la meilleure d'entre elle avec ses paramètres.

## Conclusion

A la fin, l'utilisateur a donc un jeu de données (brut ou plot en 2D selon API/site) qui permet de mettre en avant l'intérêt de UMAP + clustering, avec des paramètres optimaux pour qu'il puisse reproduire la chose. Ca mettrait en avant la contribution principale du papier initial, et en plus ça fait de jolis graphiques à visualiser pour le site.

**Comment by MatheoLeroyEnsae** [comment]
Ce plan me convient très bien. Merci d’avoir rappelé la logique initial du projet. 

Juste Un point d’attention pour être sûr :

- je comprends donc que l’objectif final est donc d optimiser la silhouette et construire une pipeline complète ? (et non la trustworthiness même si elle sera facile à monitorer) plutôt que d’optimiser en deux temps ? 

Merci pour la clarification !

**Comment by slithiaote** [comment]
Pour moi, l'esprit du cours est de mettre un service ou un outil à disposition d'un utilisateur lambda. L'utilisateur n'est pas expert et veut juste un truc qui marche. On peut faire un calcul compliqué (cela n'intéresse pas l'utilisateur), mais la réponse doit être simple du point de vue de l'utilisateur, fiable, facilement utilisable dans ses propres travaux. 

Un autre aspect de la "mise en production" est de faire évoluer le système, ce que l'on ne fait pas (encore) dans ce projet. Cela correspondrait à quelque chose comme ça : 
- on met un jeu de paramètres v0.1 dans mlflow. 
- on met un bouton "J'aime" / "J'aime pas" dans Streamlit pour avoir le feedback de l'utisateur.
- on monitore le feedback. (chacun d'entre nous envoie quelques jeux de données et fait un feedback)
- (optionnel) on fait des "runs" d'entraînement avec des logs dans mlflow
- on fait évoluer le jeu de paramètres avec une v0.2 en utilisant le "best model" de mlflow et un déploiement automatique par un push sur l'autre repo, comme dans le cours.

**Comment by mateomorin** [comment]
Conclusion : objectif trop ambitieux car explore pleinement les voies MLOps et Application, autant rester modeste et garantir une stabilité de l'ensemble des produits proposés (ainsi qu'une réponse plus rapide de l'API).

On va donc rester sur la mise en place d'outils plus simples à gérer et qui ne nécessitent pas une refonte de code entière pour assouvir un besoin spécifique d'optimisation. On laisse l'utilisateur optimiser par lui-même via streamlit ou API.

Le commentaire de @slithiaote est repris dans l'issue #24 et tant mieux : c'est intéressant et applicable pour nous.
